### PR TITLE
feat: support cloning buffers

### PR DESCRIPTION
**Pull request by salmanm**
Fixes https://github.com/davidmarkclements/rfdc/issues/18

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -80,99 +108,115 @@ test('circles and proto option – deep circular array', async ({same, is, isNot
...
+    isNot(clonedBuffer, inputBuffer, 'cloned buffer is not same as input buffer')
+    isNot(clonedBuffer.toString, input, 'cloned buffer content is correct')
+  })
+  test(`${label} copies buffers from arrays correctly`, async ({ ok, is, isNot }) => {
+    const input = Date.now().toString(36)
+    const inputBuffer = Buffer.from(input)
+    const [clonedBuffer] = clone([inputBuffer])
+    ok(Buffer.isBuffer(clonedBuffer), 'cloned value is buffer')
+    isNot(clonedBuffer, inputBuffer, 'cloned buffer is not same as input buffer')
+    isNot(clonedBuffer.toString, input, 'cloned buffer content is correct')
```
Comment (feugy) [inline test/index.js:220]: This assertion is wrong: you're not actually testing the buffer content.

uh uh

```js
is(clonedBuffer.toString(), input, 'cloned buffer content is correct')
```

will make it.

File: test/index.js [hunk 2]
```diff
@@ -80,99 +108,115 @@ test('circles and proto option – deep circular array', async ({same, is, isNot
...
     }
     fn(1, 2, 3)
   })
+  test(`${label} copies buffers from object correctly`, async ({ ok, is, isNot }) => {
+    const input = Date.now().toString(36)
+    const inputBuffer = Buffer.from(input)
+    const clonedBuffer = clone({ a: inputBuffer }).a
+    ok(Buffer.isBuffer(clonedBuffer), 'cloned value is buffer')
+    isNot(clonedBuffer, inputBuffer, 'cloned buffer is not same as input buffer')
+    isNot(clonedBuffer.toString, input, 'cloned buffer content is correct')
```
Comment (feugy) [inline test/index.js:212]: same here.

**Review by feugy — CHANGES_REQUESTED** [review]
It works really fine.

There is one case I'm wondering about: ` rfdc(Buffer.from('whatever')`.

You could inspire from how dates are handled:
```js
  function cloneProto (o) {
    if (typeof o !== 'object' || o === null) return o
    if (o instanceof Date) return new Date(o)
    // maybe something like: if (Buffer.isBuffer(o)) return Buffer.from(cur)
```

File: test/index.js (hunk 1 above)
Comment (salmanm) [reply test/index.js]: Oops! my bad.

**Comment by salmanm** [comment]
Thanks @feugy. 

I thought about `clone(Buffer.from('whatever')` but couldn't decide because while recursing, there is no way this function would be called with a buffer. So adding an extra check in `clone` (which is a recursive function) **only** to support this case could be an overhead while recursing.

I understand it's debatable, perhaps we need a third opinion. 🙂

**Comment by simoneb** [comment]
@salmanm @feugy I think they point is more about whether cloning something that is not an object, meaning the argument that's provided to the main exported `clone` function, is supported already. If it is, I believe the operation mentioned by @feugy should be supported, but I'm not sure if that's how it currently works or if you can only provide an object as input (and a few other types including dates).

@salmanm in this PR you'll also need to change README to mention that Buffers are now supported.

**Comment by salmanm** [comment]
> whether cloning something that is not an object is already supported

As per README, the function signature says `clone(obj) => obj2`. Does that count as only objects can be cloned? 😉

**Comment by simoneb** [comment]
> > whether cloning something that is not an object is already supported
> 
> As per README, the function signature says `clone(obj) => obj2`. Does that count as only objects can be cloned? 😉

It does not in my opinion, because even though that's what the readme says, the code says otherwise.

**Comment by salmanm** [comment]
Quick question for maintainers: The README shows how `Uint8Array` get's messed up while cloning. Do you see any reason why I should not fix that too as part of this PR?

**Review by simoneb — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by davidmarkclements** [comment]
Great efforts here everyone - @salmanm 
would you be okay to extend this for all binary types. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays

For best speed I’d check if there’s a buffer property that’s an instance of ArrayBuffer (`o.buffer instanceof ArrayBuffer`) and then use the slice method on the buffer property to clone the array buffer - then pass that into the constructor method of the thing being cloned (`new o.constructor(o.buffer.slice()`)

Something like that with some tweaks should work for all typed array views but node Buffers still need to be special cased (new Buffer is deprecated) so I’m fine to keep the current Buffer.from code as is and then do the type array checks under that

**Comment by salmanm** [comment]
Updated the PR @davidmarkclements.
Used `ArrayBuffer.isView` instead of checking `o.buffer instanceof`. Let me know if it may have any consequence I missed.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by salmanm** [comment]
Added a complex case of cloning ArrayBuffers in https://github.com/davidmarkclements/rfdc/pull/19/commits/20cab98a7b5a42dc4ca27c33a2cffcb2af149d16

**Comment by salmanm** [comment]
@davidmarkclements could you please take a look?

**Comment by davidmarkclements** [comment]
awesome work @salmanm I'm good with the approach, it looks good. One final thing: can you please update the readme to reflect this change

**Comment by salmanm** [comment]
@davidmarkclements updated the doc.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by davidmarkclements** [comment]
great job @salmanm - thank you!

**Pull request merged**